streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -163,7 +172,8 @@ public KafkaStreams(TopologyBuilder builder, StreamsConfig config, KafkaClientSu
             this.threads[i] = new StreamThread(builder, config, clientSupplier, applicationId, clientId, processId, metrics, time);
             storeProviders.add(new StreamThreadStateStoreProvider(threads[i]));
         }
-
+        kafkaStreamsInstances = new StreamsMetadataState(builder);
+        threads[0].setPartitionsByHostStateChangeListener(kafkaStreamsInstances);
```

Comment:
Yes i can do that. Changes are already `synchronized` anyway, so no threading issues